Python/optimizer_analysis.c

```diff
@@ -9,10 +20,608 @@
+#include "pycore_optimizer.h"
+#include "pycore_object.h"
+#include "pycore_dict.h"
+#include "pycore_function.h"
+#include "pycore_uop_metadata.h"
+#include "pycore_uop_ids.h"
+#include "pycore_range.h"
+
+#include <stdarg.h>
 #include <stdbool.h>
 #include <stdint.h>
 #include <stddef.h>
-#include "pycore_optimizer.h"
+
+// Holds locals, stack, locals, stack ... co_consts (in that order)
+#define MAX_ABSTRACT_INTERP_SIZE 4096
+
+#define OVERALLOCATE_FACTOR 5
+
+#define TY_ARENA_SIZE (UOP_MAX_TRACE_WORKING_LENGTH * OVERALLOCATE_FACTOR)
+
+// Need extras for root frame and for overflow frame (see TRACE_STACK_PUSH())
+#define MAX_ABSTRACT_FRAME_DEPTH (TRACE_STACK_SIZE + 2)
+
+#ifdef Py_DEBUG
+    static const char *const DEBUG_ENV = "PYTHON_OPT_DEBUG";
+    static inline int get_lltrace(void) {
+        char *uop_debug = Py_GETENV(DEBUG_ENV);
+        int lltrace = 0;
+        if (uop_debug != NULL && *uop_debug >= '0') {
+            lltrace = *uop_debug - '0';  // TODO: Parse an int and all that
+        }
+        return lltrace;
+    }
+    #define DPRINTF(level, ...) \
+    if (get_lltrace() >= (level)) { printf(__VA_ARGS__); }
+#else
+    #define DPRINTF(level, ...)
+#endif
+
+// See the interpreter DSL in ./Tools/cases_generator/interpreter_definition.md for what these correspond to.
+typedef enum {
+    // Types with refinement info
+    GUARD_KEYS_VERSION_TYPE = 0,
+    GUARD_TYPE_VERSION_TYPE = 1,
+    // You might think this actually needs to encode oparg
+    // info as well, see _CHECK_FUNCTION_EXACT_ARGS.
+    // However, since oparg is tied to code object is tied to function version,
+    // it should be safe if function version matches.
+    PYFUNCTION_TYPE_VERSION_TYPE = 2,
+
+    // Types without refinement info
+    PYLONG_TYPE = 3,
+    PYFLOAT_TYPE = 4,
+    PYUNICODE_TYPE = 5,
+    NULL_TYPE = 6,
+    PYMETHOD_TYPE = 7,
+    GUARD_DORV_VALUES_TYPE = 8,
+    // Can't statically determine if self or null.
+    SELF_OR_NULL = 9,
+
+    // Represents something from LOAD_CONST which is truly constant.
+    TRUE_CONST = 30,
+    INVALID_TYPE = 31,
+} _Py_UOpsSymExprTypeEnum;
+
+#define MAX_TYPE_WITH_REFINEMENT PYFUNCTION_TYPE_VERSION_TYPE
+
+static const uint32_t IMMUTABLES =
+    (
+        1 << NULL_TYPE |
+        1 << PYLONG_TYPE |
+        1 << PYFLOAT_TYPE |
+        1 << PYUNICODE_TYPE |
+        1 << SELF_OR_NULL |
+        1 << TRUE_CONST
+    );
+
+typedef struct {
+    // bitmask of types
+    uint32_t types;
+    // refinement data for the types
+    uint64_t refinement[MAX_TYPE_WITH_REFINEMENT + 1];
+    // constant propagated value (might be NULL)
+    PyObject *const_val;
+} _Py_UOpsSymType;
+
+
+typedef struct _Py_UOpsAbstractFrame {
+    // Symbolic version of co_consts
+    int sym_consts_len;
+    _Py_UOpsSymType **sym_consts;
+    // Max stacklen
+    int stack_len;
+    int locals_len;
+
+    _Py_UOpsSymType **stack_pointer;
+    _Py_UOpsSymType **stack;
+    _Py_UOpsSymType **locals;
+} _Py_UOpsAbstractFrame;
```

Comment:
I initially did that, but this new layout is easier for inlining, because it actually models the state if all frames were to be inlined.

Reply:
But we aren't inlining yet, and we might choose a different approach. Making it a single struct would make memory management simpler.

Reply:
Right now it makes memory management a little more complex, because my "frames" are just an array of static sized structs in an array. Their localsplus is just accessing a giant pool. So I just need to manage memory once instead of for 7 giant frames.